### ISSUE TITLE
【AIコース】大課題 毒キノコの判定 SVM

### ISSUE THREAD

**Issue by hayyoshi** [hayyoshi/DIC#4]
### SVMとは何か
# どのようなときに使うのか
# 分類と回帰の両方で使える手法。
# どのような仕組みなのか
# 分類(線形)
#  SVMは、どの点にも被らないように、できるだけ幅の広い帯(境界線)を引きます。
#  仕組みとしては、帯(境界線)の事をマージンと呼び、SVMはマージンを最大化し、その真ん中が判別のための境界線になります。
#  なお、最も近い点のことをサポートベクターと呼びます。
# 分類(非線形)
#  カーネルを用いて高次元空間へと移し分類を行なう。
# 回帰
#  サンプルをある距離εのマージンの中にうまく含むようにする。これをイプシロンチューブと呼び
#  その外のサンプルに対しては遠いほど一次関数的に増加するペナルティを課します。
#  ある線から離れるほど駄目というのではなく、一定の範囲（εチューブの中）は誤差がないものとして扱う。
#  ペナルティの増加は一次関数になっているので、外れ値の影響は小さくなっています。
# どういった利点欠点があるのか
# 利点として、2クラス分類だけでなく、多クラス分類が行なえる。また線形だけでなく、カーネルを使うことで非線形な分類も可能な点が挙げられる。。
# 欠点としては、以下の2点が挙げられる。
#  1対他(one-versus-the-rest,ovr)の場合、各SVMにおいて２つのクラスのバランスが悪く、また、各SVMによる予測に差が出てそれぞれの優劣を決められない。
#  1対1(one-versus-one,ovo)の場合、n*(n-1)/2の回数のSVMを行う必要があり、時間がかかりすぎる。

### 必要なライブラリをimport
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
from IPython.display import display

### データを取得する
df = pd.read_csv('agaricus-lepiota.data',header=None)
df.head(10)

### データの変換
from sklearn import preprocessing

# https://github.com/hlin117/data-science/blob/master/classification/mushroom.py
target_columns = ['target','1. cap-shape','2. cap-surface','3. cap-color','4. bruises?','5. odor','6. gill-attachment',
                  '7. gill-spacing','8. gill-size','9. gill-color','10. stalk-shape','11. stalk-root',
                  '12. stalk-surface-above-ring','13. stalk-surface-below-ring','14. stalk-color-above-ring',
                  '15. stalk-color-below-ring','16. veil-type','17. veil-color','18. ring-number',
                  '19. ring-type','20. spore-print-color','21. population','22. habitat:']
df.columns = target_columns
    
for i in target_columns:
    le = preprocessing.LabelEncoder()#データを数値で表現する
    le = le.fit(df[i])
    df[i] = le.transform(df[i])

### データセットの分割
from sklearn.model_selection import train_test_split
X = df.drop('target',axis=1)
y = df['target']
X_train,X_test,y_train,y_test =train_test_split(X,y,test_size=0.2,random_state=0)

### 標準化
# https://diver.diveintocode.jp/affiliations/textbooks/509
from sklearn.preprocessing import StandardScaler
stdsc = StandardScaler()
# 訓練用のデータのみの平均、標準偏差を用いて標準化する
X_train_std = stdsc.fit_transform(X_train)
# テストデータも標準化
X_test_std = stdsc.transform(X_test)
print(X_train_std.mean())
print(X_train_std.std())
print(X_test_std.mean())
print(X_test_std.std())

# なぜSVMを行う際に標準化が必要なのか説明してください。
# 特徴量間でスケールが異なる事で、デルをうまく学習させることができないことがある。
# その為、学習前に特徴量間でスケールを揃える必要がある。

### カーネルとは何か
# 非線形の場合、カーネルを用いて高次元空間へと移し、複雑な分類への対応を行う。
# カーネルには以下の種類がある。カーネルによって
# linear（線形カーネル）
# poly（多項式カーネル）
# rbf（RBFカーネル）
# sigmoid（シグモイドカーネル）
# precomputed（事前に計算したものを利用する）
# カーネルによって分類の仕方が異なる為、データに応じて使い分ける。
# 例えば、線形で分類可能なのであれば、線形カーネルを用いる。
# https://diver.diveintocode.jp/affiliations/textbooks/566

### コストペナルティCとは何か
# 誤分類に対するコストペナルティの値。コストぺなるティが大きいほど、過学習が発生しやすくなる。

### ハイパーパラメータを調整する
from sklearn.grid_search import GridSearchCV
from sklearn.svm import SVC
from sklearn.metrics import classification_report
from sklearn.metrics import r2_score, make_scorer

#http://scikit-learn.org/stable/modules/grid_search.html
param_grid = [
    {'C': [1,10,100,1000], 'kernel': ['linear']},
    {'C': [1, 10, 100, 1000], 'gamma': [0.001, 0.0001], 'kernel': ['rbf']},
]
scoring_fnc=make_scorer(r2_score)
# https://diver.diveintocode.jp/affiliations/textbooks/531
grid = GridSearchCV(SVC(),param_grid,scoring=scoring_fnc,cv=10)

#グリッドサーチ実施
grid.fit(X_train,y_train) #完了まで時間を要する。

### SVMの実行
#グリッドサーチによるパラメーター一覧を確認
print(grid.grid_scores_)
#グリッドサーチによるベストパラメーターを確認
print(grid.best_params_) # {'C': 1000, 'gamma': 0.001, 'kernel': 'rbf'}
#グリッドサーチによるベストパラメーターを指定したSVMによる学習
rbf_svm = SVC(kernel='rbf', gamma=0.001, C=1000, random_state=0)
rbf_svm.fit(X_train,y_train)
#グリッドサーチによるベストパラメーターを指定したSVMによる学習のスコア確認
from sklearn.metrics import accuracy_score
predictions = rbf_svm.predict(X_test)
print(accuracy_score(y_test,predictions)) # 1.0
